pkg/metrics/neutron.go

```diff
@@ -0,0 +1,84 @@
+// SPDX-License-Identifier: MIT
+
+package metrics
+
+import (
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/layer3/floatingips"
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/klog"
+)
+
+var (
+	neutronFloatingIPStatus    *prometheus.GaugeVec
+	neutronFloatingIPCreated   *prometheus.GaugeVec
+	neutronFloatingIPUpdatedAt *prometheus.GaugeVec
+
+	floatingIPLabels = []string{"id", "floating_ip", "fixed_ip", "port_id"}
+)
+
+func registerNeutronMetrics() {
+	neutronFloatingIPStatus = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: generateName("neutron_floating_ip_status"),
+			Help: "Neutron floating ip status",
+		},
+		floatingIPLabels,
+	)
+	neutronFloatingIPCreated = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: generateName("neutron_floatingip_created_at"),
+			Help: "Neutron floating ip created at",
+		},
+		floatingIPLabels,
+	)
+	neutronFloatingIPUpdatedAt = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: generateName("neutron_floatingip_updated_at"),
+			Help: "Neutron floating ip updated at",
+		},
+		floatingIPLabels,
+	)
+
+	prometheus.MustRegister(neutronFloatingIPStatus)
+	prometheus.MustRegister(neutronFloatingIPCreated)
+	prometheus.MustRegister(neutronFloatingIPUpdatedAt)
+}
+
+// ScrapeNeutronMetrics makes the list request to the neutron api and passes
+// the result to a scrape function.
+func ScrapeNeutronMetrics(neutronClient *gophercloud.ServiceClient, tenantID string) error {
+	// first step: gather the data
+	mc := newOpenStackMetric("floating_ip", "list")
+	pages, err := floatingips.List(neutronClient, floatingips.ListOpts{}).AllPages()
+	if mc.Observe(err) != nil {
+		// only warn, maybe the next scrape will work.
+		klog.Warningf("Unable to scrape floating ips: %v", err)
+		return err
+	}
+	floatingIPList, err := floatingips.ExtractFloatingIPs(pages)
+	if err != nil {
+		// only warn, maybe the next scrape will work.
+		klog.Warningf("Unable to scrape floating ips: %v", err)
+		return err
+	}
+
+	// second step: reset the old metrics
+	neutronFloatingIPStatus.Reset()
+
+	// third step: publish the metrics
+	for _, fip := range floatingIPList {
+		scrapeFloatingIPMetric(fip)
```

Comment:
I know we used it somewhere else, but scrape is usually what Prometheus is doing when retrieving the metrics from the /metrics endpoint. So I think scrape is probably the wrong word here and a bit confusing.